fendermint/vm/topdown/src/observe.rs

```diff
@@ -0,0 +1,240 @@
+        = register_int_gauge!("topdown_parent_finality_voting_latest_sent_height", "Parent finality peer");
+    TOPDOWN_PARENT_FINALITY_VOTING_QUORUM_HEIGHT: IntGauge
+        = register_int_gauge!(
+            "topdown_parent_finality_voting_quorum_height",
+            "Parent finality vote tally new agreement; recorded whenever the latest epoch with quorum"
+        );
+    TOPDOWN_PARENT_FINALITY_VOTING_QUORUM_WEIGHT: IntGauge
+        = register_int_gauge!(
+            "topdown_parent_finality_voting_quorum_weight",
+            "Parent finality vote tally new agreement; recorded whenever the latest epoch with quorum"
+        );
+    TOPDOWN_PARENT_FINALITY_COMMITTED_HEIGHT: IntGauge
+        = register_int_gauge!("topdown_parent_finality_committed_height", "Parent finality committed on chain");
+}
+
+impl_traceables!(
+    TraceLevel::Info,
+    "Topdown",
+    ParentRpcCalled<'a>,
+    ParentFinalityAcquired<'a>,
+    ParentFinalityPeerVoteReceived<'a>,
+    ParentFinalityPeerVoteSent<'a>,
+    ParentFinalityPeerQuorumReached<'a>,
+    ParentFinalityCommitted<'a>
+);
+
+#[derive(Debug)]
+pub struct ParentRpcCalled<'a> {
+    pub source: &'a str,
+    pub json_rpc: &'a str,
+    pub method: &'a str,
+    pub status: &'a str,
+    pub latency: f64,
+}
+
+impl Recordable for ParentRpcCalled<'_> {
+    fn record_metrics(&self) {
+        TOPDOWN_PARENT_RPC_CALL_TOTAL
+            .with_label_values(&[self.source, self.method, self.status])
+            .inc();
+
+        TOPDOWN_PARENT_RPC_CALL_LATENCY_SECS
+            .with_label_values(&[self.source, self.method, self.status])
+            .observe(self.latency);
+    }
+}
+
+pub type BlockHeight = u64;
+/// Hex encoded block hash.
+pub type BlockHashHex<'a> = &'a str;
+
+#[derive(Debug)]
+pub struct ParentFinalityAcquired<'a> {
+    pub source: &'a str,
+    pub is_null: bool,
+    pub block_height: BlockHeight,
+    pub block_hash: Option<BlockHashHex<'a>>,
+    pub commitment_hash: Option<BlockHashHex<'a>>,
+    pub num_msgs: usize,
+    pub num_validator_changes: usize,
+}
+
+impl Recordable for ParentFinalityAcquired<'_> {
+    fn record_metrics(&self) {
+        TOPDOWN_PARENT_FINALITY_LATEST_ACQUIRED_HEIGHT
+            .with_label_values(&[self.source])
+            .set(self.block_height as i64);
+    }
+}
+
+#[derive(Debug)]
+pub struct ParentFinalityPeerVoteReceived<'a> {
+    pub validator: &'a str,
+    pub block_height: BlockHeight,
+    pub block_hash: BlockHashHex<'a>,
+    pub commitment_hash: BlockHashHex<'a>,
+}
+
+impl Recordable for ParentFinalityPeerVoteReceived<'_> {
+    fn record_metrics(&self) {
+        TOPDOWN_PARENT_FINALITY_VOTING_LATEST_RECEIVED_HEIGHT
+            .with_label_values(&[self.validator])
+            .set(self.block_height as i64);
+    }
+}
+
+#[derive(Debug)]
+pub struct ParentFinalityPeerVoteSent<'a> {
+    pub block_height: BlockHeight,
+    pub block_hash: BlockHashHex<'a>,
```

Comment:
why keep these two fields if they are not used?

Reply:
They will be stored in the tracing event right.